Tasks/MSBuild/msbuild.ts

```diff
@@ -0,0 +1,124 @@
+
+//read inputs
+var solution = tl.getPathInput('solution', true, false);
+var platform = tl.getInput('platform');
+var configuration = tl.getInput('configuration');
+var msbuildArguments = tl.getInput('msbuildArguments');
+var clean = tl.getBoolInput('clean');
+var logsolutionEvents = tl.getBoolInput('logsolutionEvents');
+if(logsolutionEvents) {
+    tl.warning('logSolutionEvents property is not supported on the VSTS node agent.');
+}
+var msbuildLocationMethod = tl.getInput('msbuildLocationMethod');
+if(!msbuildLocationMethod) {
+    msbuildLocationMethod = 'version';
+}
+
+var xbuildToolPath = tl.which('xbuild'); //ignore msbuild version on non-Windows platforms, use xbuild
+if(msbuildLocationMethod == 'location') {
+    xbuildToolPath = tl.getInput('msbuildLocation');
+}
+
+var runxbuild = function (fn, clean) {
+    return Q.fcall( () => {
+        var xbuild = tl.createToolRunner(xbuildToolPath);
+        xbuild.arg(fn);
+        if(clean) {
+            xbuild.arg('/t:Clean');
+        }
+        if (platform) {
+            xbuild.arg('/p:Platform=\"' + platform + '\"');
+        }
+        if (configuration) {
+            xbuild.arg('/p:Configuration=\"' + configuration + '\"');
+        }
+        if (msbuildArguments) {
+            xbuild.arg(msbuildArguments);
+        }
+
+        return xbuild.exec();
+    });
+}
+
+// Resolve files for the specified value or pattern
+var filesList : string [];
+if (solution.indexOf('*') == -1 && solution.indexOf('?') == -1) {
+    // No pattern found, check literal path to a single file
+    tl.checkPath(solution, 'files');
+
+    // Use the specified single file
+    filesList = [solution];
+
+} else {
+    var firstWildcardIndex = function(str) {
+        var idx = str.indexOf('*');
+
+        var idxOfWildcard = str.indexOf('?');
+        if (idxOfWildcard > -1) {
+            return (idx > -1) ?
+                Math.min(idx, idxOfWildcard) : idxOfWildcard;
+        }
+
+        return idx;
+    }
+
+    // Find app files matching the specified pattern
+    tl.debug('Matching glob pattern: ' + solution);
+
+    // First find the most complete path without any matching patterns
+    var idx = firstWildcardIndex(solution);
+    tl.debug('Index of first wildcard: ' + idx);
+    var findPathRoot = path.dirname(solution.slice(0, idx));
+
+    tl.debug('find root dir: ' + findPathRoot);
+
+    // Now we get a list of all files under this root
+    var allFiles = tl.find(findPathRoot);
+
+    // Now matching the pattern against all files
+    filesList = tl.match(allFiles, solution, {matchBase: true});
+
+    // Fail if no matching .csproj files were found
+    if (!filesList || filesList.length == 0) {
+        tl.error('No matching files were found with search pattern: ' + solution);
+        tl.exit(1);
+    }
+}
+
+var result = Q({});
+if(clean) {
+    filesList.forEach((fn) => {
+        result = result.then(() => {
+            return runxbuild(fn, true);
```

Comment:
I will rearrange this to be a little more clear but we need to run xbuild twice in this scenario. Good catch on tl.exit(0) in the first then, it works now since tl.exit(0) actually does not exit.